mlir/lib/Dialect/Vector/Transforms/LowerVectorGather.cpp

```diff
@@ -96,6 +96,82 @@ struct FlattenGather : OpRewritePattern<vector::GatherOp> {
   }
 };
 
+/// Rewrites a vector.gather of a strided MemRef as a gather of a non-strided
+/// MemRef with updated indices that model the strided access.
+///
+/// ```mlir
+/// %subview = memref.subview %M (...) to memref<100xf32, strided<[3]>>
+/// %gather = vector.gather %subview (...) : memref<100xf32, strided<[3]>>
+/// ```
+/// ==>
+/// ```mlir
+/// %collapse_shape = memref.collapse_shape %M (...) into memref<300xf32>
+/// %1 = arith.muli %idxs, %c3 : vector<4xindex>
+/// %gather = vector.gather %collapse_shape (...) : memref<300xf32> (...)
+/// ```
+///
+/// ATM this is effectively limited to reading a 1D Vector from a 2D MemRef,
+/// but should be fairly straightforward to extend beyond that.
+struct RemoveStrideFromGatherSource : OpRewritePattern<vector::GatherOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(vector::GatherOp op,
+                                PatternRewriter &rewriter) const override {
+    Value base = op.getBase();
+    if (!base.getDefiningOp())
+      return failure();
+
+    // TODO: Strided accesses might be coming from other ops as well
+    auto subview = dyn_cast<memref::SubViewOp>(base.getDefiningOp());
```

Comment:
Based on the comment, have you considered to use `OffsetSizeAndStrideOpInterface` interface?

btw, we prefer moving the casting to `getDefiningOp` method, e.g.,

```suggestion
    auto subview = base.getDefiningOp<memref::SubViewOp>();
```

Reply:
> btw, we prefer moving the casting to getDefiningOp method

Thanks! Would you have a link to a relevant discussion? Otherwise I will probably repeat similar mistake :(

Reply:
There's also a subtle (safety) difference, `base.getDefiningOp()` can return null, which will fail to `dyn_cast`. `getDefiningOp<memref::SubViewOp>()` accounts for that (via doing a `dyn_cast_or_null` internally).

Reply:
I actually don't have a pointer to a relevant discussion. It is mostly from my review experience and codebase exploring. It is shorter and safer, so I think it is a hidden convention.